core/src/main/java/org/kiwix/kiwixmobile/core/main/TabsAdapter.kt

```diff
@@ -0,0 +1,170 @@
+      RecyclerView.LayoutParams.WRAP_CONTENT,
+      RecyclerView.LayoutParams.MATCH_PARENT
+    )
+    val textView = TextView(context)
+    textView.id = 4
+    textView.maxLines = 1
+    textView.ellipsize = TextUtils.TruncateAt.END
+    constraintLayout.addView(
+      textView,
+      ConstraintLayout.LayoutParams(0, ConstraintLayout.LayoutParams.WRAP_CONTENT)
+    )
+    val constraintSet = ConstraintSet()
+    constraintSet.clone(constraintLayout)
+    constraintSet.connect(
+      cardView.id, ConstraintSet.TOP, ConstraintSet.PARENT_ID,
+      ConstraintSet.TOP
+    )
+    constraintSet.connect(
+      cardView.id, ConstraintSet.BOTTOM, ConstraintSet.PARENT_ID,
+      ConstraintSet.BOTTOM
+    )
+    constraintSet.connect(
+      cardView.id, ConstraintSet.START, ConstraintSet.PARENT_ID,
+      ConstraintSet.START, margin16
+    )
+    constraintSet.connect(
+      cardView.id, ConstraintSet.END, ConstraintSet.PARENT_ID,
+      ConstraintSet.END, margin16
+    )
+    constraintSet.connect(close.id, ConstraintSet.END, cardView.id, ConstraintSet.END)
+    constraintSet.connect(close.id, ConstraintSet.BOTTOM, cardView.id, ConstraintSet.TOP)
+    constraintSet.connect(
+      textView.id, ConstraintSet.BOTTOM, cardView.id,
+      ConstraintSet.TOP
+    )
+    constraintSet.connect(
+      textView.id, ConstraintSet.START, cardView.id,
+      ConstraintSet.START, margin16 / 8
+    )
+    constraintSet.connect(textView.id, ConstraintSet.END, close.id, ConstraintSet.START)
+    constraintSet.applyTo(constraintLayout)
+    return ViewHolder(constraintLayout, contentImage, textView, close)
+  }
+
+  override fun onBindViewHolder(holder: ViewHolder, position: Int) {
+    val webView = webViews[position]
+    webView.parent?.let { (it as ViewGroup).removeView(webView) }
+    val webViewTitle = webView.title.fromHtml().toString()
+    holder.title.text = webViewTitle
+    holder.close.setOnClickListener { v: View -> listener?.onCloseTab(v, holder.adapterPosition) }
+    holder.content.setImageBitmap(
+      getBitmapFromView(webView, activity.getWindowWidth(), activity.getWindowHeight())
+    )
+    holder.content.setOnClickListener { v: View ->
+      selected = holder.adapterPosition
+      listener?.onSelectTab(v, selected)
+      notifyDataSetChanged()
+    }
+    if (webViewTitle != activity.getString(R.string.menu_home)) {
+      painter.update(holder.content)
+    }
+  }
+
+  override fun getItemCount(): Int = webViews.size
+
+  override fun getItemId(position: Int): Long = webViews[position].hashCode().toLong()
+
+  fun setTabClickListener(listener: TabClickListener) {
+    this.listener = listener
+  }
+
+  interface TabClickListener {
+    fun onSelectTab(view: View, position: Int)
+    fun onCloseTab(view: View, position: Int)
+  }
+
+  class ViewHolder(view: View, val content: ImageView, val title: TextView, val close: ImageView) :
+    RecyclerView.ViewHolder(view)
+
+  init {
```

Comment:
should be at the top

Reply:
https://github.com/kiwix/kiwix-android/pull/2452/commits/ac979a2534bac443b4fb2dc2f10a4542d994f5b9